Remove settings json construct from ENV

freddydk opened the pull request:
Fix for issue #620

Things changed in this PR:
- ReadSettings will now output the settings to GITHUB_OUTPUT instead of GITHUB_ENV
- Actions expecting a settingsJSON will no longer set the input parameter to ENV:_settingsJson, causing the PS script to have an encoding problem (settings might contain special characters) - therefore ReadSettings now needs to output as Base64
- ReadSettings has a parameter on whether to use base64 - set to Y in this version, notifying the action that Base64 response is expected. If this was not done - we might get issues during Update AL-Go System files.
- ReadSecrets and DetermineArtifactUrl did earlier just re-write the settings environment variable. With GITHUB_OUTPUT, they now have to output to GITHUB_OUTPUT as well and the subsequent usages of settings needs to be from there
- ReadSecrets and DetermineArtifactUrl will output base64 if they were given base64 - only ReadSettings will have the useBase64 parameter.

Additional fix:
- removed the string around $ENV:_parameter in the sign action.yaml - no reason for that.

If you want to try this out - you can update AL-Go System Files with freddydk/AL-Go@SettingsEnv - until the branch is deleted (after the merge of this PR)

mazhelez commented on Actions/DetermineArtifactUrl/action.yaml:
```diff
@@ -22,6 +22,9 @@ inputs:
     required: false
     default: '{"insiderSasToken":""}'
 outputs:
+  SettingsJson:
```
Why would this action output settings?

freddydk replied:
DetermineArtifactUrl updates the settings (in the "old" days it re-wrote the ENV:SETTINGS)
We could instead return the artifactURL, but then we would have to do more changes to CI/CD in order to get this artifact into the settings - or use it from outside the setting.

I didn't want to do major restructuring - instead I was aiming at creating the least disruptive change for removing the ENV problem.

mazhelez commented on Actions/CalculateArtifactNames/CalculateArtifactNames.ps1:
```diff
@@ -19,8 +19,13 @@ function Set-EnvVariable([string] $name, [string] $value) {
-Write-Host $settingsJson
-$settings = $settingsJson | ConvertFrom-Json
+# Support potentially base64 encoded settings
+try {
```
This construct seems off. Shouldn't the action's contract specify if the settings are base64-encoded or not?

freddydk replied:
Due to the nature of our preview - we will have old versions of AL-Go workflows calling into this function with a json construct which isn't base64 - and every new version will be base64.
I assume that once we have shipped this - we will be able to remove the support for non-base64 in preview - but for now, we have to support both in order to not break partners.

I could also have added the useBase64 parameter - but that one (as this construct) is temporary and can be removed later.

mazhelez commented on Actions/ReadSettings/ReadSettings.ps1:
```diff
@@ -87,18 +89,23 @@ try {
-    Add-Content -Path $env:GITHUB_ENV -Value "Settings=$OutSettingsJson"
-    Write-Host "SettingsJson=$outSettingsJson"
+    if ($useBase64) {
+        Add-Content -Path $env:GITHUB_OUTPUT -Value "SettingsJson=$([Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes($outSettingsJson)))"
```
Readability: Move `Add-Content -Path $env:GITHUB_OUTPUT -Value "SettingsJson=$SettingsJsonValue"` outside of the if-else.

freddydk replied:
fixed

mazhelez commented on Actions/ReadSettings/ReadSettings.ps1:
```diff
@@ -87,18 +89,23 @@ try {
-    Add-Content -Path $env:GITHUB_ENV -Value "Settings=$OutSettingsJson"
-    Write-Host "SettingsJson=$outSettingsJson"
+    if ($useBase64) {
+        Add-Content -Path $env:GITHUB_OUTPUT -Value "SettingsJson=$([Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes($outSettingsJson)))"
```
Isn't the actual problem that we output every setting (line 85 in this file: `Add-Content -Path $env:GITHUB_ENV -Value "$setting=$($settingValue | ConvertTo-Json -Depth 99 -Compress)`).

So, with that change, every setting will be in `$env:GITHUB_ENV` and all the settings (JSON-formatted and maybe base64-encoded) will the in the `$env:GITHUB_OUTPUT`. 

I'm not convinced the issue will be solved.

freddydk replied:
The real problem is that settings where as individual values + as a json construct 2 or 3 times.
This PR removes the json constructs.
It might be a good idea to generally look at what ENV variables are needed - but for now, this fixes the issue (I am convinced:-)) and we can fix the other ones later,

freddydk replied:
This is a build of ALAppExtensions with the new ENV mechanism:

mazhelez commented on Tests/TestActionsHelper.psm1:
```diff
@@ -83,21 +83,29 @@ function YamlTest {
-                $envLines.AppendLine("        _$($name): `${{ inputs.$($name) }}")
-                if ($type -eq "System.String" -or $type -eq "System.Int32") {
-                    $parameterString += " -$($name) `$ENV:_$($name)"
+                if ($name -eq 'settingsJson') {
```
A comment in here would be useful to understand why the special case for `settingsJson`.

mazhelez commented on Actions/ReadSecrets/ReadSecrets.ps1:
```diff
@@ -113,12 +123,17 @@ try {
+    $outSettingsJson = $outSettings | ConvertTo-Json -Depth 99 -Compress
+    Write-Host "- SettingsJson=$outSettingsJson"
+    if ($useBase64) {
+        $outSettingsJson = [Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes($outSettingsJson))
```
Since at this point the `$outSettingsJson` may contain secrets, it's quite dangerous to encode it as the encoded string is in no way masked (see [here](https://github.com/freddydk/ALAppExtensions/actions/runs/5657371091/job/15327074618#step:8:100) for example).

freddydk replied:
Adding the entire settings construct will mask the encoded settings and minimize the clutter in the logs

mazhelez reviewed (changes requested):
Settings may contain secrets so encoding them is a security concern. This needs to be addressed.

freddydk commented on Actions/CalculateArtifactNames/CalculateArtifactNames.ps1:
```diff
@@ -19,8 +19,13 @@ function Set-EnvVariable([string] $name, [string] $value) {
-$settings = $settingsJson | ConvertFrom-Json
+# Support potentially base64 encoded settings
+try {
+    $settings = [System.Text.Encoding]::UTF8.GetString([System.Convert]::FromBase64String($settingsJson)) | ConvertFrom-Json
```
Yes, ConvertTo-HashTable is a ContainerHelper function and ContainerHelper is not loaded here.
settings will then be a PSCustomObject

freddydk commented on Actions/ReadSecrets/action.yaml:
```diff
@@ -16,16 +16,20 @@ inputs:
     default: '7b7d'
+outputs:
+  SettingsJson:
+    description: Settings in compressed Json format
```
Or, ReadSecrets and DetermineArtifactUrl should not re-write the entire settings construct.

freddydk commented:
In order to test that the action.yaml matches the DetermineArtifactUrl.ps1, I have renamed the DetermineArtifactUrl.Action.ps1 to DetermineArtifactUrl and added tests to call YamlTest (which builds the yaml based on the .ps1 and checks that they are identical)
